Feature: Add custom serial buttons extension

sevimuelli opened the pull request:
Adds a new extension for creating buttons to send custom commands.

Addresses discussion [101](https://github.com/jetkvm/kvm/discussions/101), [134](https://github.com/jetkvm/kvm/discussions/134), Pull request [324](https://github.com/jetkvm/kvm/pull/324) and Issue [267](https://github.com/jetkvm/kvm/issues/267).

Mostly used to control an external KVM so one can switch ports inside the JetKVM, but I tried to keep it as universal as possible. Would probably also work with your Serial Console Adapter. Everything is tested and all automated test passed. As it looks like many people are using the serial port to control or connect external devices, this extension could help a lot of users. Here are some screenshots:

Basic view:
<img width="454" height="373" alt="Screenshot 2025-09-23 at 23 03 15" src="https://github.com/user-attachments/assets/196da0db-d487-4078-82ae-ab3af1757786" />

Serial settings menu:
<img width="439" height="638" alt="Screenshot 2025-09-23 at 23 03 27" src="https://github.com/user-attachments/assets/4f7004f7-8970-4286-ad87-4dc237d8cf8d" />

Button config menu:
<img width="440" height="573" alt="Screenshot 2025-09-23 at 23 03 45" src="https://github.com/user-attachments/assets/9caef25d-3e71-4de2-a015-70d7a62acca8" />

Response view:
<img width="441" height="550" alt="Screenshot 2025-09-23 at 23 04 11" src="https://github.com/user-attachments/assets/d7423b3f-3bdb-46c8-a755-5f5e65fea5ae" />

sevimuelli committed:
Create new extension "Serial Buttons"

sevimuelli committed:
Merge branch 'jetkvm:dev' into dev

sevimuelli committed:
Add backend to send custom commands

sevimuelli committed:
Add order buttons and response field

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/jetkvm/kvm?pullRequest=821) <br/>All committers have signed the CLA.

IDisposable commented:
Code review pending, but one design suggestion. The line-ending should probably be set in each step (not globally) and be `None`, `CR \r`, `LF \n`, or `CR/LF \r\n`... that way each button can send _exactly_ as desired instead of forcing a line-ending to be sent and forcing them all to be the same.

IDisposable commented on jsonrpc.go:
```diff
@@ -893,6 +906,64 @@ func rpcSetSerialSettings(settings SerialSettings) error {
+
+type SerialButtonConfig struct {
+	Buttons            []QuickButton `json:"buttons"`            // slice of QuickButton
+	Terminator         string        `json:"terminator"`         // CR/CRLF/None
```
Move `Terminator` to the `QuickButton` it is settable per-button.

IDisposable commented on jsonrpc.go:
```diff
@@ -893,6 +906,64 @@ func rpcSetSerialSettings(settings SerialSettings) error {
+}
+
+func rpcSetSerialButtonConfig(config SerialButtonConfig) error {
+
```
_nit_ delete this blank line (codebase pattern)

IDisposable commented on jsonrpc.go:
```diff
@@ -893,6 +906,64 @@ func rpcSetSerialSettings(settings SerialSettings) error {
+		HideSerialResponse: true,
+	}
+
+	file, err := os.Open("/userdata/serialButtons_config.json")
```
Since this path is mentioned several places in this file, perhaps extract to a file constant?

IDisposable commented on serial.go:
```diff
@@ -251,6 +252,83 @@ func setDCRestoreState(state int) error {
+					time.Sleep(50 * time.Millisecond)
+					continue
+				}
+				if n == 0 || currentSession == nil {
```
Do we want to drain the serial port if the `currentSession` is `nil`? Seems like the check for nil and `continue` should be up before line 288 (e.g. before reading the port) and might be worth adding a sleep to that clause.

IDisposable commented on serial.go:
```diff
@@ -251,6 +252,83 @@ func setDCRestoreState(state int) error {
+					continue
+				}
+				// Safe for any bytes: wrap in Base64
+				b64 := base64.StdEncoding.EncodeToString(buf[:n])
```
Not sure we should be manually base-64 encoding, the RPC channel already knows how to transport raw `byte` arrays (see keyboard stuff).

sevimuelli replied:
My fought was that raw strings only work if the things comming over serial is UTF-8 text (ASCII commands, logs, etc.). But let's say we get binary control bytes (0x00, 0xFF, etc.) or text in a non-UTF-8 encoding, this could produce invalide UTF-8. Idk i i'm wrong and if it makes sense. What do you think? I can easily remove it.

IDisposable replied:
I see what you're getting at, since this is the data coming back to the browser from the device (from the serial device), so yes it could easily be any arbitrary sequence of bytes... but that's already fine in an RPC message. Even "key down state" messages could have any byte sequences and they transport fine... look at the embedded go routine in serial.go `handleSerialChannel`... it's just emitting the bytes right into the buffer.

(and if you end up merging the extensions, you will want to be able to support the existing window)

sevimuelli replied:
Ahh yess, then i will remove that. Just wanted to be save. Thanks for clarifying.

IDisposable commented on serial.go:
```diff
@@ -251,6 +252,83 @@ func setDCRestoreState(state int) error {
+	// Stop previous loop if running
+	if serialButtonsRXStopCh != nil {
+		close(serialButtonsRXStopCh)
+	}
```
Make this call to the `stopSerialButtonsRxLoop` function.

IDisposable commented on serial.go:
```diff
@@ -251,6 +252,83 @@ func setDCRestoreState(state int) error {
+	}()
+}
+
+func stopSerialButtonsRxLoop() {
```
If you pass in the scopedLogger, you could debug-log the close attempt.

IDisposable commented on serial.go:
```diff
@@ -251,6 +252,83 @@ func setDCRestoreState(state int) error {
+func sendCustomCommand(command string) error {
+	scopedLogger := serialLogger.With().Str("service", "custom_buttons_tx").Logger()
+	scopedLogger.Info().Str("Command", command).Msg("Sending custom command.")
+	_, err := port.Write([]byte("\n"))
```
Shouldn't this "\n" be selected by the desired line-ending in settings (or step/modal). A case could be made that if we're asked to send a custom command, we shouldn't be tacking on the line-ending, we should send it verbatim.

IDisposable commented on serial.go:
```diff
@@ -251,6 +252,83 @@ func setDCRestoreState(state int) error {
+	scopedLogger.Info().Str("Command", command).Msg("Sending custom command.")
+	_, err := port.Write([]byte("\n"))
+	if err != nil {
+		scopedLogger.Warn().Err(err).Msg("Failed to send serial output \\n")
```
The `\\n` is to emit to the logger the line-ending (which should be coming from settings/modal) and could easily be confusing in the logs. I suggest just not  adding that to the message.

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+
+/** ============== Types ============== */
+
+interface SerialSettings {
```
This seems to be in conflict/duplication with the settings in the current serial extension. To that point, perhaps this code should be used in place of the current serial extension (e.g. add the Open Console here, delete the old extension code), or see if it is possible to extract and reuse the port-settings to a shared component?

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+
+interface ButtonConfig {
+  buttons: QuickButton[];
+  terminator: string;   // CR/CRLF/None
```
Should allow LF only and LFCR too...

IDisposable replied:
Please consider moving the terminator selection to each individual button.

sevimuelli replied:
Good point, done!

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+    if (!chunk) return;
+
+    // Normalize CRLF for display
+    chunk = chunk.replace(/\r\n/g, "\n");
```
This normalization seems likely to cause much [confusion and delay](https://www.youtube.com/watch?v=PuADGZBIg88). 
Perhaps we should consider replacing all control characters with their ASCII "label" (and tacking on the literal `'\n'` after the selected terminator ... e.g. something like what [grok suggests](https://grok.com/c/8d0de4ff-daef-46e3-8f0d-1328e3646ff6)

```typescript
function formatControlChars(input: string): string {
  const asciiControlCharMap: { [key: number]: string } = {
    0: '<NUL>', 1: '<SOH>', 2: '<STX>', 3: '<ETX>', 4: '<EOT>',
    5: '<ENQ>', 6: '<ACK>', 7: '<BEL>', 8: '<BS>', 9: '<TAB>',
    10: '<LF>', 11: '<VT>', 12: '<FF>', 13: '<CR>', 14: '<SO>',
    15: '<SI>', 16: '<DLE>', 17: '<DC1>', 18: '<DC2>', 19: '<DC3>',
    20: '<DC4>', 21: '<NAK>', 22: '<SYN>', 23: '<ETB>', 24: '<CAN>',
    25: '<EM>', 26: '<SUB>', 27: '<ESC>', 28: '<FS>', 29: '<GS>',
    30: '<RS>', 31: '<US>', 127: '<DEL>'
  };

  let result = '';
  let i = 0;

  while (i < input.length) {
    const char = input[i];
    const code = char.charCodeAt(0);

    // Ensure byte range (0–255)
    if (code > 255) {
      throw new Error(`Character at index ${i} has code ${code}, which exceeds byte range (0–255)`);
    }

    // Handle space
    if (code === 0x20) {
      result += ' ';
      i++;
      continue;
    }

    // Handle ASCII control characters (0–31, 127)
    if ((code <= 31 || code === 127) && asciiControlCharMap[code]) {
      const isCR = code === 13;
      const isLF = code === 10;
      result += asciiControlCharMap[code];

      // Check for <CR><LF> or <LF><CR> sequences
      if (i + 1 < input.length) {
        const nextCode = input[i + 1].charCodeAt(0);
        if (nextCode <= 255 && ((isCR && nextCode === 10) || (isLF && nextCode === 13))) {
          result += asciiControlCharMap[nextCode];
          i += 2;
          result += '\n';
          continue;
        }
      }

      // Add newline after standalone <CR> or <LF>
      if (isCR || isLF) {
        result += '\n';
      }
      i++;
      continue;
    }

    // Handle non-ASCII bytes (128–255)
    if (code >= 128) {
      result += `<${code.toString(16).toUpperCase().padStart(2, '0')}>`;
      i++;
      continue;
    }

    // Handle printable ASCII characters (32–126, excluding space)
    result += char;
    i++;
  }

  return result;
}
```

Which would work like this...
```typescript
const input = "He\r\nll \0\x80\xff";
console.log(formatControlChars(input));
// Output: "He<CR><LF>\nll <NUL><80><FF>"

try {
  const invalid = "He\u0100"; // Code point > 255
  console.log(formatControlChars(invalid));
} catch (e) {
  console.log(e.message); // "Character at index 2 has code 256, which exceeds byte range (0–255)"
}
```

sevimuelli replied:
I like that idea. I could more or less copy paste that and use it as a helper function. Cool idea!

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+     */
+    const raw = btn.command;
+    const t = buttonConfig.terminator ?? "";
+    const command = raw.endsWith("\r") || raw.endsWith("\n") ? raw : raw + t;
```
This is why I suggest making the button own the terminator. Then you can just always emit it (and if it's empty adds nothing). As this code is currently written, there's a forced termination, no way to send without, and also if the command ends with either terminator, we ignore the set termination... so if that was set to <CR><LF> and the command string ended with \r, then we _would not_ send the correct terminator sequence.

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+
+  const saveDraft = () => {
+    const label = draftLabel.trim() || "Unnamed";
+    const command = draftCmd.trim();
```
What if I want to send a command that ends or starts with a space? I think this `.trim` should not be done.

sevimuelli replied:
Good point. Removed

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+      return;
+    }
+
+    const isEdit = editorOpen?.id;
```
s/isEdit/editId ?

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+                  options={[
+                    { label: "None", value: "" },
+                    { label: "CR (\\r)", value: "\r" },
+                    { label: "CRLF (\\r\\n)", value: "\r\n" },
```
Add
```typescript
{ label: "LF (\\n)", value: "\n" },
{ label: "LFCR (\\n\\r)", value: "\n\r" },

```

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+          {/* Serial settings (collapsible) */}
+          {!buttonConfig.hideSerialSettings && (
+            <>
+              <div className="grid grid-cols-2 gap-4">
```
This section (if we moved terminator to each button as previously suggested), could then be identical to that of the existing serial port extension, and thus could/should be extracted out to common popovers.

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+  };
+
+  /** simple reordering: alphabetical by sort, then label */
+  const sortedButtons = useMemo(() => stableSort(buttonConfig.buttons), [buttonConfig.buttons]);
```
Seems like _all_ the paths that mutate the buttonConfig end with `handleSerialButtonConfigChange("buttons", stableSort(nextButtons));` so the buttons will already/always be sorted. We probably should just bind to `buttonConfig.buttons` from here on down, and don't need to re-sort.

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+            <>
+              <hr className="border-slate-700/30 dark:border-slate-600/30" />
+              <TextAreaWithLabel
+                ref={taRef}
```
Do we need to put in the "ignore keystrokes" logic like in [PasteModal.tsx](https://github.com/jetkvm/kvm/blob/dev/ui/src/components/popovers/PasteModal.tsx#L138-L139)? I think we might want to done it that way there so that the textbox is not `readOnly` so it is not "grayed out" and thus more readable.

sevimuelli replied:
I don't think so. As you can see in the screenshot i attached, it isn't greyed out. The only thing is that you can't write in it but the text is still normal and you can also select it. It would be different with "disabled".

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+                placeholder="Will show the response recieved from the serial port."
+              />
+              <div className="flex items-center gap-2">
+                <Button
```
This really doesn't "pause" anything, it just ignores the serial data coming back from the device... what's the use-case that would drive that? It's more like a mute button for thing incoming serial...right?

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+/** ============== helpers ============== */
+
+function pretty(s: string) {
+  return s.replace(/\r/g, "\\r").replace(/\n/g, "\\n");
```
Elsewhere (in the menus, for example), we us the format `CR \\r` ... in fact, can't we just lookup the label from the list of terminator options?

sevimuelli replied:
Done. Instead of just a string, i used an object to store label and value of the selected terminator. 
```jsx
terminator: {label: string, value: string};
```

IDisposable commented on ui/src/components/popovers/ExtensionPopover.tsx:
```diff
@@ -76,6 +83,8 @@ export default function ExtensionPopover() {
         return <DCPowerControl />;
       case "serial-console":
```
We could easily merge these two extensions with a bit more work. @adamshiervani any thoughts on going that route?

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -0,0 +1,494 @@
+    // Normalize CRLF for display
+    chunk = chunk.replace(/\r\n/g, "\n");
+
+    setSerialResponse(prev => (prev + chunk).slice(-MAX_CHARS));
```
We could _actually_ pause with some logic here that enqueues the chunks (and defer calling `setSerialResponse`) when paused. Then when unpaused "play-back" the queued chunks to get to the final `setSerialResponse` call

sevimuelli commented:
> Code review pending, but one design suggestion. The line-ending should probably be set in each step (not globally) and be `None`, `CR \r`, `LF \n`, or `CR/LF \r\n`... that way each button can send _exactly_ as desired instead of forcing a line-ending to be sent and forcing them all to be the same.

Yes, I can do that change. Especially to include separate CR and LF. The reason why I made the setting global and not for each button individually is because it is highly unlikely if not impossible to have more than one client device. And I believed that one device always uses the same settings. But it's an easy change so I will adapt it. 👌

sevimuelli committed:
Merge branch 'jetkvm:dev' into feat/custom-serial-buttons

IDisposable commented:
> The reason why I made the setting global and not for each button individually is because it is highly unlikely if not impossible to have more than one client device.

Indeed if what you're doing it emitting to a serial-controlled KVM... but this extension is *SO MUCH* more useful than just that... _any_ serial connection could be driven, so very easily could create macro/buttons for a whole bunch of common commands or snippets that you might want to emit... like "start a command", [user then types into the terminal], "emit some config stuff" , [user types some more, "end the command setup" (which would have the terminator selected to something)

sevimuelli commented:
I completely agree. I think the best step would be to merge the two extensions and create a propper, highly functional and universal serial extension. I already did most of the suggested changes but some are still missing. 

To-Do:

- [ ] Remove base64
- [ ] Implement formalization (formatControlChars)
- [ ] Implement proper pause functionality

One thing I noticed about the serial console extension, I don't know if it's a ME problem or a general problem, is that if I click `open console`, I can click in the console but can't write anything. Is there maybe a problem that the port is not available? Looking at serial.go, it should be ok. I need to do some more troubleshooting. 

I'm thinking about how to best combine the buttons and the serial console. Like if you press a button to send a command, should the send command also appear in the console, should the response appear in the response window i created or in the console or both, or don't we even need my response window? I agree that the best step is to combine the two extensions but I want everything to work properly together and avoid redundant functionality or things working against each other.

IDisposable commented:
> One thing I noticed about the serial console extension, I don't know if it's a ME problem or a general problem, is that if I click `open console`, I can click in the console but can't write anything. Is there maybe a problem that the port is not available? Looking at serial.go, it should be ok. I need to do some more troubleshooting.

To be clear here... when you are in the serial console, the keys you press are _sent_ out the serial point so unless the remote echoes them back, you will see NOTHING.

What is shown in the serial console is what the remote device is sending... it could have absolutely nothing to do with what you're typing, if anything... 

Imagine that your serial port is connected to an traffic signal control port, you might just get `EASTBOUND YELLOW WESTBOUND YELLOW EASTBOUND RED WESTBOUNT LEFT GREEN` etc... and if you pressed the `N` key that might trigger right-of-way process to switch to Northbound getting the green :)

IDisposable commented:
> I'm thinking about how to best combine the buttons and the serial console. Like if you press a button to send a command, should the send command also appear in the console, should the response appear in the response window i created or in the console or both, or don't we even need my response window?

Assuming there are two distinct regions, pressing key OR pressing a macro button should send the bytes out the serial port (and show them in the "SENT" region).

The "RECV" console region should _only_ show what is received from the attached serial device. If it chooses to echo back what was sent to it... ONLY THEN would the console "RECV" region show anything, and what it show could be completely distinct from what the device _sent_.

sevimuelli commented:
First, thanks for clarifying. Now i get it! 😅
About the distinct send and receive region, what if we would only use one window, the existing serial console we already have? One could have an option for "local echo" or something, like many terminals or serial apps support it, that the things you write are echoed back into the terminal. I think this would also make it easier to follow the flow of signals instead of two distinct regions. 

So I'm proposing  to following: 
- Merge the two extensions
- Modify the existing serial window, have an option to echo commands that are typed or sent by buttons
- Remove my "response window", as it is redundant
- Still implement formalization
- Implement a proper pause functionality, the way you proposed it. So messages don't get lost but temporarily stored in a buffer

What do you think about that? Maybe you have other great improvements I could implement while I'm on it?

IDisposable commented:
Perfect! Just a little checkbox for the local echo

sevimuelli committed:
Merge extensions "Serial Console" and "Serial Buttons"

sevimuelli committed:
Merge branch 'dev' into feat/custom-serial-buttons

IDisposable commented on serial.go:
```diff
@@ -307,23 +314,25 @@ func startSerialButtonsRxLoop(session *Session) {
 	scopedLogger := serialLogger.With().Str("service", "custom_buttons_tx").Logger()
 	scopedLogger.Info().Str("Command", command).Msg("Sending custom command.")
-	_, err := port.Write([]byte("\n"))
+	_, err := port.Write([]byte(terminator))
```
Two questions:

- Do we want to wrap the emission of the terminator in an `if` block checking if it's set to something (not nil, not empty)
- I would expect the order to be send-command then send-terminator

sevimuelli replied:
Already changed

IDisposable commented on serial.go:
```diff
@@ -336,6 +345,149 @@ var defaultMode = &serial.Mode{
+
+func getSerialSettings() (CustomButtonSettings, error) {
+	config := CustomButtonSettings{
+		BaudRate:           strconv.Itoa(defaultMode.BaudRate),
```
Wouldn't it be simpler to store/use the numeric value as integers (baud,data-bits) floats (stop bits)?

sevimuelli replied:
For baudrate and databits, yes and i changed that now. for parity and stopbits, no, because inside the actual serial port, it's handled differently. For example tha parity: the different settings are associated to numbers 0 to 4. If we use that directly, it might be confusing for anyone else trying to work on that.

IDisposable commented on ui/src/components/Terminal.tsx:
```diff
@@ -161,6 +170,11 @@ function Terminal({
 
+  const sendLine = useCallback((line: string) => {
+    // Just send; echo/normalization handled elsewhere as you planned
+    dataChannel.send(line + "\r\n"); // adjust CR/LF to taste
```
... hmmm, we'll need to expose that choice somehow... but for now, it should be just `"\r"` because that's what an Enter key means in most ever serial setup (though it might be echoed back as CRLF)

sevimuelli replied:
I'm handling that in the backend in serial.go. You will see it in the next commit.

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -8,38 +8,43 @@ import { JsonRpcResponse, useJsonRpc } from "@/hooks/useJsonRpc";
-import { TextAreaWithLabel } from "@components/TextArea";
+import { useUiStore } from "@/hooks/stores";
+
+import Checkbox from "../../components/Checkbox";
```
`"@components/Checkbox"`

IDisposable commented on ui/src/components/extensions/SerialButtons.tsx:
```diff
@@ -111,70 +98,50 @@ export function SerialButtons() {
-    const raw = btn.command;
-    const t = buttonConfig.terminator ?? "";
-    const command = raw.endsWith("\r") || raw.endsWith("\n") ? raw : raw + t;
+    const command = btn.command + btn.terminator.value;
```
Won't this send the terminator twice?

IDisposable reviewed (commented):
Looking much nicer. Are you still planning on Play/Pause ?

sevimuelli commented:
Still a work in progress. Frontend is finished more or less. Now working on properly combining the data from the serial terminal and the buttons and routing all traffic including local echo to the console, with normalization. Just wanted to make a commit for the progress so far. But I was waking up with my Laptop still on top of me in the middle of the night, so...

I will finish the backend and then make a commit. About the pause functionality, I'm not sure if it's actually needed. I was planning on omitting it. What do you think?

IDisposable commented:
I think we can ship without pause, it's a nice to have, but not required.

sevimuelli committed:
Update backend to combine serial console and custom buttons

sevimuelli committed:
Merge branch 'jetkvm:dev' into feat/custom-serial-buttons

sevimuelli commented:
Still a work in progress. I just had to commit to update some things. Final version should be done later today

sevimuelli commented:
Not directly related but I can't continue developing because the web interface is stuck after updating the branch.  I created a issue https://github.com/jetkvm/kvm/issues/870. Any ideas? Any help is welcome.

sevimuelli committed:
Merge branch 'jetkvm:dev' into feat/custom-serial-buttons

sevimuelli committed:
Update backend, implement pause function in terminal

sevimuelli committed:
Merge branch 'jetkvm:dev' into feat/custom-serial-buttons

sevimuelli committed:
Improve normalization

sevimuelli commented:
Now everything looks fine and seems to work. I added some options to configure normalization and also added a pause button in the terminal. The last thing I want to do is some real world testing, connecting my USB-to-RS232 adapter and simulating some things. 

The only thing I looked at but couldn't figure out how to fix it is the issue with the focus of the terminal. When you open the terminal and then open a popover, you can't use the terminal anymore and have to close and reopen it. It has to do with this focus trap. Would be amazing if this can be fixed.

IDisposable commented:
Sounds like the modal is NOT disabling the keyboard trap.

sevimuelli committed:
Minor serial helper improvements

sevimuelli commented:
So I tested everything. Did some extensive testing with my serial adapter and also my 10-port KVM that a lot of people here seem to use. Made some improvements. Now everything works and from my side, is ready to be merged.

About the other issue, I wound out that is has to do with disableVideoFocusTrap, which pauses FocusTrap. When you open the terminal, FocusTrap is momentarily paused. Then you can click between the terminal and WebRTCVideo to type. But once you open a popup, it's not possible anymore to do anything with the terminal. You have to close it and reopen it again. I tried some things but didn't get it working. I think there are a couple things one can do to fix it, but that's probably something for another PR or issue, maybe next time I need something to do 😉

sevimuelli committed:
Merge branch 'dev' into feat/custom-serial-buttons

sevimuelli commented:
Hi, would it be possible to do the final checks and merge it if everything is fine?

IDisposable commented on serial.go:
```diff
@@ -368,62 +316,139 @@ type QuickButton struct {
+
+	var normalizeMode NormalizeMode
+	switch serialConfig.NormalizeMode {
+	case "carret":
```
s/carret/caret

IDisposable commented on serial.go:
```diff
@@ -368,62 +316,139 @@ type QuickButton struct {
+		crlfMode = CRLF_CR
+	case "crlf":
+		crlfMode = CRLF_CRLF
+	case "lfcr":
```
This case is not listed in the comment https://github.com/jetkvm/kvm/pull/821/files/4ddce3f0ee20b452ffcd07e6b24d78f2cafdd249..7c09ac3c08645fa56a48a6595609caa43b0008f3#diff-ad4fb56affb62e9f5024c32ea8a857376661813e1bc22b7dfef64f21ff16c34fR328

sevimuelli replied:
I added it. Even when it's non standard and hardly ever showes up, i wanted to include every edge case.

IDisposable commented on serial.go:
```diff
@@ -474,15 +499,56 @@ func setSerialSettings(newSettings CustomButtonSettings) error {
+
+	var normalizeMode NormalizeMode
+	switch serialConfig.NormalizeMode {
+	case "carret":
```
s/carret/caret

IDisposable commented on serial.go:
```diff
@@ -510,80 +576,61 @@ func reopenSerialPort() error {
-		// }
+	dataChannel.OnMessage(func(msg webrtc.DataChannelMessage) {
+
+		scopedLogger.Info().Bytes("Data:", msg.Data).Msg("Sending data to serial mux")
```
Seems like duplicate logging.

IDisposable commented on serial_console_helpers.go:
```diff
@@ -87,34 +89,68 @@ func normalize(in []byte, opt NormOptions) string {
+				} else {
+					i++
+				}
+				out.WriteString("\r\n") // (fixed to actually write CRLF)
```
drop the AI comment ;)

IDisposable commented on serial_console_helpers.go:
```diff
@@ -184,11 +220,22 @@ type ConsoleBroker struct {
+	quietAfter   time.Duration
 
 	// normalization
 	norm NormOptions
```
Maybe better to spell out the type to `NormalizationOptions`

sevimuelli replied:
Agree!

IDisposable commented on serial_console_helpers.go:
```diff
@@ -200,42 +247,78 @@ type ConsoleBroker struct {
+		sink:           s,
+		in:             make(chan consoleEvent, 256),
+		done:           make(chan struct{}),
+		pauseCh:        make(chan bool, 8),
```
Why would we have more than... 2 (pause/unpause)...

sevimuelli replied:
In order to avoid race conditions. Also, with this aproach, you know when you unpause the broker, so you can flush the buffered input.

IDisposable commented on serial_console_helpers.go:
```diff
@@ -268,16 +351,49 @@ func (b *ConsoleBroker) stopQuietTimer() {
+		}
+
+		// Update line-end state based on this piece
+		atLineEnd = endsWithEOL(line, b.norm.CRLF)
```
This name is a bit confusing... would `w.norm.LineEnding` be clearer?

IDisposable commented on serial_console_helpers.go:
```diff
@@ -312,6 +451,122 @@ func (b *ConsoleBroker) flushPendingTX() {
+	var parts []string
+	start := 0
+	for i := 0; i < len(b); i++ {
+		if b[i] == '\r' || b[i] == '\n' {
```
This little snippet occurs twice (in almost identical form) and it complex enough that it could easily drift... can we pull this out to a distinct helper function? See similar code at https://github.com/jetkvm/kvm/pull/821/files/4ddce3f0ee20b452ffcd07e6b24d78f2cafdd249..7c09ac3c08645fa56a48a6595609caa43b0008f3#diff-4ab4a20836f8a4f9e8bfcf3796e5ce01291b59439e00547fc75240a531b35c38R93-R113

sevimuelli replied:
I see the similarities but I don't see a good way to create a helper function because the both do quite different things inside. What do you think?

IDisposable commented on serial_console_helpers.go:
```diff
@@ -312,6 +451,122 @@ func (b *ConsoleBroker) flushPendingTX() {
+	b.bufLines = append(b.bufLines, s)
+	b.bufBytes += len(s)
+	// trim if over limits (drop oldest)
+	for b.bufBytes > b.maxBufBytes || len(b.bufLines) > b.maxBufLines {
```
Nice :)

IDisposable commented on serial_console_helpers.go:
```diff
@@ -312,6 +451,122 @@ func (b *ConsoleBroker) flushPendingTX() {
+
+func (b *ConsoleBroker) flushBuffer() {
+	if b.sink == nil || len(b.bufLines) == 0 {
+		b.bufLines = nil
```
Will setting this to `nil` break the appends later?

sevimuelli replied:
No, this isn't a problem. nil is just an empty, uninizilaized slice. You can later append to nil safely, no proplem.

IDisposable commented on serial_console_helpers.go:
```diff
@@ -370,24 +627,28 @@ func (m *SerialMux) reader() {
 			}
 			// echo (if requested AND globally enabled)
 			if f.echo && m.echoEnabled.Load() && m.broker != nil {
+				scopedLogger.Info().Msg("Sending TX echo to console broker")
```
Almost all (if not all) of these `.Info()` should really be `.Trace()` level... this is _really_ chatty

IDisposable commented:
This is _really_ close... just a few polishes needed

sevimuelli committed:
Update serial console part

sevimuelli committed:
Merge branch 'jetkvm:dev' into feat/custom-serial-buttons

sevimuelli commented:
So I did everything except the one to extract the snippet into a helper function.

sevimuelli committed:
Small bug fixes

sevimuelli commented:
So I noticed that you added localization, should I add that everywhere to the new SerialConsole, so it can finally be merged? And if so, could you tell me where I need to enter the translation?

IDisposable commented:
You can just merge dev into your branch and I or Adam or Avi can do the base translation...or feel free to see the section on Localization in the DEVELOPMENT.MD file 😁

sevimuelli committed:
Merge branch 'dev' into feat/custom-serial-buttons

sevimuelli committed:
Add localization

sevimuelli commented:
I added all the necessary localization. The only thing that I wasn't able to do is the translation with `npm run i18n:machine-translate`. Could you merge it and then do the translation?

sevimuelli committed:
Merge branch 'jetkvm:dev' into feat/custom-serial-buttons

sevimuelli committed:
Merge branch 'dev' into feat/custom-serial-buttons

sevimuelli committed:
Fix linting error

sevimuelli commented:
Hi, i'm currently unable to test things on a real device because i'm traveling. Would it please be possible that you review this pull request and merge it so I don't have to keep updating it? Would be great if it is included in the next release. Thanks 😊

itsMaxC commented:
This feature is fantastic, thanks for implementing it! Hopefully it gets released soon.

sevimuelli commented:
Hi @IDisposable 
Any idea when we can merge that pull request? It's been sitting here for a while.

IDisposable commented:
> Hi @IDisposable Any idea when we can merge that pull request? It's been sitting here for a while.

I don't generally merge PRs, just a contributor.

itsMaxC commented:
@adamshiervani I'm taking the liberty of tagging you here to attract your attention to this PR, since this seem to be a prime candidate for merging and, shamelessly, something that would help my workflow. My apologies if I'm out of line.

sevimuelli commented:
> > Hi @IDisposable Any idea when we can merge that pull request? It's been sitting here for a while.
> 
> 
> 
> I don't generally merge PRs, just a contributor. 

Hi, I'm wondering what I would need to do to finally merge this. It's been finished and sitting around for months now 🥺

sevimuelli commented:
I guess not. Would have preferred though that you write in your documentation that contributions will be ignored. Could have saved me many hours. @IDisposable @adamshiervani

adamshiervani commented:
Hey @sevimuelli,

Sorry for the latency on merging this! Could you just resolve the three merge conflicts, and i'll merge as soon as that is done?

Thanks!

sevimuelli committed:
Merge branch 'dev' into feat/custom-serial-buttons

sevimuelli commented:
Hi @adamshiervani 

No worries, I understand. I just didn't understand why it was sitting there for so long. I resolved the conflicts and updated the branch. Now everything should be ready to merge. As I mentioned in an earlier comment, I can't test the updates on a physical device because I'm traveling and for some reason it doesn't work through Tailscale. Just to keep in mind. But as I always tested everything, I expect it to work. Also, from an earlier comment:

> I added all the necessary localization. The only thing that I wasn't able to do is the translation with `npm run i18n:machine-translate`. Could you merge it and then do the translation?

Thank you. I'm excited to see it in action!

sevimuelli committed:
Fix UI linting issues

sevimuelli committed:
Fix Go lint issues

IDisposable reviewed (approved):
:shipit:

sevimuelli committed:
Differentiate between User and System commands

sevimuelli commented:
I made an slight improvement again. It's only a beauty correction but now things that get send from the system (terminal size), are separately marked and also, after this message, the terminal automatically starts on a newline.

I did the ui and golang lint as well as testing on my own device and everything works as expected.

Now I won't make changes anymore, I promise ;)

sevimuelli commented:
Thank you 🙏. I'm excited to see it in the next release!
